File: dotnet/base/Mcma.Core/Logging/Logger.cs
using System;

namespace Mcma.Logging
{
    public abstract class Logger : ILogger
    {
        protected Logger(string source, string requestId, McmaTracker tracker)
        {
            Source = source;
            RequestId = requestId ?? Guid.NewGuid().ToString();
            Tracker = tracker;
        }

        protected string Source { get; }

        protected string RequestId { get; }

        protected McmaTracker Tracker { get; }

        public static ILogger System { get; set; } = new ConsoleLogger(nameof(System));

        protected LogEvent BuildLogEvent(int level, string type, string message, object[] args)
            => new LogEvent(type, level, Source, RequestId, DateTime.UtcNow, message, args, Tracker);

        protected abstract void WriteLogEvent(LogEvent logEvent);
        
        public void Log(int level, string type, params object[] args)
            => WriteLogEvent(BuildLogEvent(level, type, null, args));
        
        public void Log(int level, string type, string message, params object[] args)
            => WriteLogEvent(BuildLogEvent(level, type, message, args));

        public void Fatal(string message, params object[] args)
            => Log(LogLevel.Fatal, LogType.Fatal, message, args);
        public void Fatal(params object[] args)
            => Log(LogLevel.Fatal, LogType.Fatal, null, args);
        
        public void Error(string message, params object[] args)
            => Log(LogLevel.Error, LogType.Error, message, args);
        public void Error(params object[] args)
            => Log(LogLevel.Error, LogType.Error, null, args);
        
        public void Warn(string message, params object[] args)
            => Log(LogLevel.Warn, LogType.Warn, message, args);
        public void Warn(params object[] args)
            => Log(LogLevel.Warn, LogType.Warn, null, args);
        
        public void Info(string message, params object[] args)
            => Log(LogLevel.Info, LogType.Info, message, args);
        public void Info(params object[] args)
            => Log(LogLevel.Info, LogType.Info, null, args);

        public void Debug(string message, params object[] args)
            => Log(LogLevel.Debug, LogType.Debug, message, args);
        public void Debug(params object[] args)
            => Log(LogLevel.Debug, LogType.Debug, null, args);

        public void FunctionStart(string message, params object[] args)
            => Log(LogLevel.FunctionEvent, LogType.FunctionStart, message, args);
        public void FunctionEnd(string message, params object[] args)
            => Log(LogLevel.FunctionEvent, LogType.FunctionEnd, message, args);

        public void JobStart(params object[] args)
            => Log(LogLevel.Info, LogType.JobStart, null, args);
        public void JobStart(string message, params object[] args)
            => Log(LogLevel.Info, LogType.JobStart, message, args);
        public void JobUpdate(params object[] args)
            => Log(LogLevel.Info, LogType.JobUpdate, null, args);
        public void JobUpdate(string message, params object[] args)
            => Log(LogLevel.Info, LogType.JobUpdate, message, args);
        public void JobEnd(params object[] args)
            => Log(LogLevel.Info, LogType.JobEnd, null, args);
        public void JobEnd(string message, params object[] args)
            => Log(LogLevel.Info, LogType.JobEnd, message, args);
    }
}

File: nodejs/base/core/lib/logging/logger.ts
import * as util from "util";
import { v4 as uuidv4 } from "uuid";

import { McmaTrackerProperties } from "../model";
import { LogEvent } from "./log-event";

export enum LogLevel {
    Fatal = 100,
    Error = 200,
    Warn = 300,
    Info = 400,
    Debug = 500
}

export enum LogType {
    Fatal = "FATAL",
    Error = "ERROR",
    Warn = "WARN",
    Info = "INFO",
    Debug = "DEBUG"
}

export interface Logger {
    debug(message: any, ...optionalParams: any[]): void;
    info(message: any, ...optionalParams: any[]): void;
    warn(message: any, ...optionalParams: any[]): void;
    error(message: any, ...optionalParams: any[]): void;
    fatal(message: any, ...optionalParams: any[]): void;

    functionStart(message: any, ...optionalParams: any[]): void;
    functionEnd(message: any, ...optionalParams: any[]): void;

    jobStart(message: any, ...optionalParams: any[]): void;
    jobUpdate(message: any, ...optionalParams: any[]): void;
    jobEnd(message: any, ...optionalParams: any[]): void;
}

export abstract class Logger implements Logger {
    private readonly _source: string;
    private readonly _requestId: string;
    private readonly _tracker?: McmaTrackerProperties;

    protected constructor(source: string, requestId?: string, tracker?: McmaTrackerProperties) {
        this._source = source;
        this._requestId = requestId ?? uuidv4();
        this._tracker = tracker;
    }

    protected get source(): string {
        return this._source;
    }
    protected get requestId(): string {
        return this._requestId;
    }
    protected get tracker(): McmaTrackerProperties {
        return this._tracker;
    }

    static System: Logger;

    protected buildLogEvent(level: number, type: string, message: any, ...optionalParams: any[]): LogEvent {
        if (optionalParams.length) {
            if (typeof message === "string") {
                message = util.format(message, ...optionalParams);
            } else {
                message = [message, ...optionalParams];
            }
        }

        return new LogEvent(type, level, this.source, this.requestId, new Date(), message, this.tracker);
    }

    fatal(message: any, ...optionalParams: any[]): void {
        this.log(LogLevel.Fatal, LogType.Fatal, message, ...optionalParams);
    }

    error(message: any, ...optionalParams: any[]): void {
        this.log(LogLevel.Error, LogType.Error, message, ...optionalParams);
    }

    warn(message: any, ...optionalParams: any[]): void {
        this.log(LogLevel.Warn, LogType.Warn, message, ...optionalParams);
    }

    info(message: any, ...optionalParams: any[]): void {
        this.log(LogLevel.Info, LogType.Info, message, ...optionalParams);
    }

    debug(message: any, ...optionalParams: any[]): void {
        this.log(LogLevel.Debug, LogType.Debug, message, ...optionalParams);
    }

    functionStart(message: any, ...optionalParams: any[]): void {
        this.log(450, "FUNCTION_START", message, ...optionalParams);
    }

    functionEnd(message: any, ...optionalParams: any[]): void {
        this.log(450, "FUNCTION_END", message, ...optionalParams);
    }

    jobStart(message: any, ...optionalParams: any[]): void {
        this.log(LogLevel.Info, "JOB_START", message, ...optionalParams);
    }

    jobUpdate(message: any, ...optionalParams: any[]): void {
        this.log(LogLevel.Info, "JOB_UPDATE", message, ...optionalParams);
    }

    jobEnd(message: any, ...optionalParams: any[]): void {
        this.log(LogLevel.Info, "JOB_END", message, ...optionalParams);
    }
    
    log(level: number, type: string, message: any, ...optionalParams: any[]): void {
        this.writeLogEvent(this.buildLogEvent(level, type, message, ...optionalParams));
    }

    protected abstract writeLogEvent(logEvent: LogEvent): void;
}
